joint_limits_controller/src/limits.cpp

```diff
@@ -0,0 +1,224 @@
+                return;
+            }
+            if(!limits_nh.getParam("k_velocity", soft_limits.k_velocity)){
+                ROS_ERROR("k_velocity not set");
+                return;
+            }
+        }
+        limits_flags |= SoftLimitsConfigured;
+
+    }
+}
+
+bool JointLimiter::Limits::getAccelerationLimit(double &limit,const double& period) const{
+    if(hasJerkLimits()){
+        limit = period * joint_limits.max_jerk;
+        if(joint_limits.has_acceleration_limits && limit > joint_limits.max_acceleration) limit = joint_limits.max_acceleration;
+    }else if(hasAccelerationLimits()){
+        limit = joint_limits.max_acceleration;
+    }else{
+        return false;
+    }
+    return true;
+}
+
+bool JointLimiter::Limits::getVelocityLimit(double &limit,const double& period) const{
+    double a;
+    if(getAccelerationLimit(a, period)){
+        limit = a*period;
+        if(joint_limits.has_velocity_limits && limit > joint_limits.max_velocity) limit = joint_limits.max_velocity;
+    } else if(hasVelocityLimits()){
+        limit = joint_limits.max_velocity;
+    }else{
+        return false;
+    }
+    return true;
+}
+
+bool JointLimiter::Limits::valid() const {
+    return true;
+}
+
+#define TRY_MERGE(flag, max_prop, has_prop, msg)                                        \
+    do {                                                                                \
+        if(other.limits_flags & flag){                                                  \
+            bool ok = true;                                                             \
+            if((limits_flags & flag ) && joint_limits.has_prop){                        \
+                if(!other.joint_limits.has_prop){                                       \
+                    ok = false;                                                         \
+                    ROS_ERROR_STREAM("Cannot disable " << msg << " limits");            \
+                }else if(other.joint_limits.max_prop > joint_limits.max_prop){          \
+                    ok = false;                                                         \
+                    ROS_ERROR_STREAM("new joint_limits.max_" << msg << " too high");    \
+                }                                                                       \
+            }                                                                           \
+            if(ok){                                                                     \
+                limits_flags |= flag;                                                   \
+                joint_limits.max_prop = other.joint_limits.max_prop;                    \
+                joint_limits.has_prop = other.joint_limits.has_prop;                    \
+            }                                                                           \
+        }                                                                               \
+    }while(0);
```

Comment:
No, the `do..while(0)` is required for function-like macros.
`{}` is ambiguous if used together with conditional clauses. 
However, the last semicolon slipped through.